File: heap_sort.py
#!/usr/bin/env python
 

def heap_sort(lst):
    def sift_down(start, end):
        """max-heapify"""
        root = start
        while True:
            child = 2 * root + 1
            if child > end:
                break
            if child + 1 <= end and lst[child] < lst[child + 1]:
                child += 1
            if lst[root] < lst[child]:
                lst[root], lst[child] = lst[child], lst[root]
                root = child
            else:
                break
 
    # build max-heap 
    for start in range(int((len(lst) - 2) / 2), -1, -1):
        sift_down(start, len(lst) - 1)
 
    # heap sort
    for end in range(len(lst) - 1, 0, -1):
        lst[0], lst[end] = lst[end], lst[0]
        sift_down(0, end - 1)
    return lst


def generate_random_integer_list(n, stop=100):
    from random import randrange
    lst = []
    for i in range(n):
        lst.append(randrange(stop))
    return lst


def main():
    num = raw_input('Input the length of list-> ')
    try:
        num = int(num)
    except ValueError:
        print "Input para error! we will run with default para."
        num = 0
    if num < 1:
        num = 20 #default para
    list1 = generate_random_integer_list(num, num * 10)
    print "Data are as follows:"
    print list1
    list2 = heap_sort(list1)
    print "After sorting: "
    print list2 

if __name__ == "__main__":
    main()

File: heap_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define NUM 10000

void show_array(int arr[], int len)
{
	int i;
	for (i = 0; i < len; i ++)
	{
		printf("%-7d", arr[i]);
		if ((i % 10) == 9) printf("\n");
	}
	printf("\n");
}

static void swap(int *a, int *b)
{
  int tmp = *a;
  *a = *b;
  *b = tmp;
}

static void create_random_array(int *arr, int n)
{
	int i;
	srand(time(0));
	for (i = 0; i < n; i ++)
		arr[i] = rand() % (n * 10);
	printf("生成的数组如下：\n");
	show_array(arr, n);
}

void max_heapify(int *arr, int idx, int len)
{
	int root = idx;
	int child = root * 2 + 1;

	while (child < len)
	{
		if (child + 1 < len && arr[child] < arr[child + 1])
			child ++;
		if (arr[root] < arr[child])
		{
			swap(&arr[root], &arr[child]);
			root = child;
			child = root * 2 + 1;
		}
		else
		{
			break;
		}
	}
}

void heap_sort(int *arr, int len)
{
	int i;

	for (i = (len - 2) / 2; i >= 0; i --)
		max_heapify(arr, i, len);
	for (i = 0; i < len; i ++)
	{
		swap(&arr[0], &arr[len - i - 1]);
		max_heapify(arr, 0, len - i - 1);
	}
}

int main(void)
{
	int a[NUM] = {0};
	int n = NUM;

	printf("Input the length(not more than %d) of array:", NUM);
	scanf("%d", &n);
	if (n > NUM || n < 1)
	{
		printf("Input para error!\n");
		return 0;
	}
	create_random_array(a, n);
	heap_sort(a, n);
	printf("\n堆排序后:\n");
	show_array(a, n);

	return 0;
}

